compiler/env/CompileTimeProfiler.hpp

```diff
@@ -0,0 +1,171 @@
+ * is met or the current compilation ends.
+ *
+ * Perf, by default, is configured to collect call graph information at a
+ * high frequency, with the following options:
+ *   --call-graph -F9999
+ * These defaults can be overridden using the TR_CompileTimeProfiler env var.
+ *
+ * The current implementation only supports perf, but other profiling
+ * utilities could be introduced. 
+ *
+ * In testing, the initial portion of the compile was lost, due to perf's
+ * start up. A delay has been added to manage this.
+ */
+class CompileTimeProfiler
+   {
+public:
+   CompileTimeProfiler(TR::Compilation *comp, const char *identifier)
+      {
+      _pid = 0;
+      if (comp->getOption(TR_CompileTimeProfiler))
+         {
+         char timestr[7];
+         time_t timer = time(NULL);
+         sprintf(timestr, "%i", (int32_t)timer % 100000);
+         timestr[6] = '\0';
+         
+         char tidstr[11];
+         snprintf(tidstr, 11, "%d", syscall(SYS_gettid));
+         tidstr[10] = '\0';
+
+         // Construct the output filename, with the thread id and time in seconds
+         char filename[1025];
+         snprintf(filename, 1025, "%.1000s.%s.%s.data", identifier ? identifier : "perf", tidstr, timestr);
+
+         // Build up the options, with the env var override
+         char *options[1025];
+         size_t pos = 0;
+         options[pos++] = "perf";
+         options[pos++] = "record";
+         options[pos++] = "-o";
+         options[pos++] = filename;
+         options[pos++] = "-t";
+         options[pos++] = tidstr;
+
+         static char *envOptions = feGetEnv("TR_CompileTimeProfiler");
+         if (envOptions)
+            {
+            char buffer[1025];
+            snprintf(buffer, 1025, "%s", envOptions);
+
+            // Replace spaces with null terminators and
+            // add words to options list
+            bool newArg = true;
+            char *iter = buffer;
+            while (*iter && pos < 1000)
+               {
+               if (*iter == ' ')
+                  {
+                  *iter = '\0';
+                  newArg = true; 
+                  }
+               else if (newArg)
+                  {
+                  options[pos++] = iter;
+                  newArg = false;
+                  }
+               iter++;
+               }
+            }
+         else
+            {
+            options[pos++] = "--call-graph";
+#if defined(TR_TARGET_X86) && defined(TR_TARGET_64BIT)
+            options[pos++] = "dwarf";
+#endif
+            options[pos++] = "-F9999";
+            }
+
+         options[pos++] = NULL;
+
+         // Dump the arguments
+         printf("\nProfiling %s compile time with:", comp->signature());
+         char **iter = options;
+         while (*iter)
+            {
+            printf(" %s", *iter);
+            iter++;
+            }
+         printf("\n");
+
+         _pid = fork();
+         if (_pid == 0)
+            {
+            execv("/usr/bin/perf", options);
```

Comment:
I haven't thought this through but I'll throw it out there while noodling it over and to invite other comments.  I'm concerned we're creating a security hole here, though how to exploit it I haven't decided.  If someone were to replace /usr/bin/perf with their own malicious application (not easy, but theoretically possible) what kind of mayhem could ensue with the privileges it is granted from the JVM process?  The JIT option that controls this is enabled in production builds (i.e., its not a debug option), hence you can force this code to run from any Linux build.  I'd like some reassurance that we're not creating an opportunity for exploitation here, and if we are we need to protect ourselves.  Thinking...

Reply:
Having said that though, in order to add this option to the command-line or JIT environment you'll have to be able to launch the runtime process anyway.  So maybe you're no worse off.  This code might be ok, I just have an uneasy feeling about it.

Reply:
I am not aware of an approach to easily make this more secure. I am happy to hide it behind a build option, if that would help.

Reply:
In order to replace `/usr/bin/perf` you'd have to acquire root or equivalent privileges, at which point you have the keys to the kingdom, so this doesn't seem like a hole in that context. Programs should be able to execute other programs, particularly those in secure locations like `/usr`, and OpenJ9 already allows you to invoke arbitrary commands via `-Xdump`.

Reply:
Considering @ymanton's contribution, this seems to be safe due to the fully specified path and the use of a command line option to enable the feature.